src/Fmt_ast.ml

```diff
@@ -303,18 +303,54 @@ let rec fmt_longident (li: Longident.t) =
       cbox 0 (fmt_longident li1 $ wrap "(" ")" (fmt_longident li2))
 
 
-let fmt_constant const =
+let fmt_constant (c: Conf.t) const =
+  let is_printable char_code = 0x20 <= char_code && char_code <= 0x7E in
+  let escape_char chr =
+    let code = Char.to_int chr in
+    match c.Conf.escape_chars with
+    | _ when Char.is_whitespace chr || Char.equal chr '\\' ->
+        str (Char.escaped chr)
+    | _ when is_printable code -> char chr
+    | `Hexadecimal -> str (Printf.sprintf "\\x%02x" code)
+    | `Minimal -> char chr
+    | `Octal -> str (Char.escaped chr)
+  in
   match const with
   | Pconst_integer (lit, suf) | Pconst_float (lit, suf) ->
       str lit $ opt suf char
-  | Pconst_char c -> wrap "'" "'" @@ str (Char.escaped c)
+  | Pconst_char '\'' -> wrap "'" "'" @@ str "\\'"
+  | Pconst_char c -> wrap "'" "'" @@ escape_char c
   | Pconst_string (s, delim) ->
+      let escape_literal string =
+        let looks_like_format =
+          String.fold string ~init:(false, true) ~f:
+            (fun (prev_is_at, acc) -> function
+            | '@' when prev_is_at -> (false, acc)
+            | '@' -> (true, acc)
+            | '\n' when prev_is_at -> (false, true && acc)
+            | '\n' -> (false, false)
+            | _ -> (false, acc) )
+          |> snd
+        in
```

Comment:
How about
```
let looks_like_format index =
  index <> 0 && Char.equal string.[index - 1] '@'
```
or am I missing something?

Reply:
Cross-posted with @hcarty

Reply:
wouldn't that bring too many "false positives"?